### PR TITLE
order book

### PR DESCRIPTION
**Pull request by RustemYuzlibaev**

### CODE REVIEW

File: packages/util/src/orderBook/index.ts [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+   *
+   */
+  public async getOrderBooks(): Promise<void> {
+    const toKey = (address) => address.code.toString();
+
+    const entries = await this.root.api.query.orderBook.orderBooks.entries();
+
+    const orderBooks = entries.reduce((buffer, [key, value]) => {
+      const orderBook = key.args[0].toJSON();
+      const meta = value.unwrap();
```
Comment (stefashkaa) [inline packages/util/src/orderBook/index.ts:27]: Error: unwrapping of none value

Please use unwrap or default with the default value you set (maybe null).

File: packages/util/src/orderBook/index.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+      const { orderBookId, status, lastOrderId, tickSize, stepLotSize, minLotSize, maxLotSize } = meta;
+      const base = orderBookId.toJSON().base;
```
Comment (stefashkaa) [inline packages/util/src/orderBook/index.ts:29]: const orderBook = orderBookId.toJSON()

please re-use it

File: packages/util/src/orderBook/index.ts [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+   * @param orderBookId base and quote addresses
+   */
+  public getAggregatedBids(base: string, quote: string): Observable<Array<[]>> {
+    return this.root.apiRx.query.orderBook.aggregatedBids({ dexId: 0, base, quote }).pipe(
+      map((data) => {
+        const priceAmountMapping = data.toJSON() as any;
+        let bids = [];
+
+        for (const bid of Object.keys(priceAmountMapping)) {
+          const price = this.getValue(JSON.parse(bid));
```
Comment (stefashkaa) [inline packages/util/src/orderBook/index.ts:112]: Why JSON.parse?

File: packages/util/src/orderBook/index.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    quote: string,
+    price: string,
+    amount: string,
+    side: Side,
+    timestamp = MAX_TIMESTAMP
+  ): Promise<T> {
+    return this.root.submitExtrinsic(
+      this.root.api.tx.orderBook.placeLimitOrder({ dexId: 0, base, quote }, price, amount, side, timestamp),
+      this.root.account.pair,
+      {} as HistoryItem
```
Comment (stefashkaa) [inline packages/util/src/orderBook/index.ts:180]: {} as HistoryItem -> maybe we need to specify it?

File: packages/util/src/orderBook/types.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+export enum OrderBookStatus {
+  // All operations are allowed
```
Comment (stefashkaa) [inline packages/util/src/orderBook/types.ts:2]: please remove whitespaces & use /** */ comments

File: packages/util/src/orderBook/index.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+import { FPNumber } from '@sora-substrate/math';
+import { HistoryItem } from '../BaseApi';
+import { LimitOrder, Side, Value } from './types';
+
+import type { Observable } from '@polkadot/types/types';
+
+export class OrderBookModule<T> {
+  constructor(private readonly root: Api<T>) {}
+
+  public orderBooks = {};
```
[the remark this answers is not in the document]
Comment (RustemYuzlibaev) [reply packages/util/src/orderBook/index.ts:13]: it should be requested only once as Order Books will not be updated too often. Then you look only into orderBooks object (no need to be refetched)

File: packages/util/src/orderBook/index.ts (hunk 3 above)
Comment (RustemYuzlibaev) [reply packages/util/src/orderBook/index.ts:112]: because general-purpose function expects this input, it's used in many other places

**Review by stefashkaa — APPROVED** [review]

**Pull request merged**